app/summary/page.tsx

```diff
@@ -0,0 +1,143 @@
+"use client";
+
+import untypedSubjectSummaries from "@/app/generated/subject-summaries.json";
+import { SummaryRow } from "@/app/types";
+import { useState, useRef, useEffect } from "react";
+
+const SUMMARY_DISPLAY_COUNT = 100;
+const subjectSummaries = untypedSubjectSummaries as unknown as SummaryRow[];
+const tableHeaders = [
+    {
+        key: "enrollmentTerm",
+        label: "Term"
+    },
+    {
+        key: "subjectArea",
+        label: "Subject"
+    },
+    {
+        key: "catalogNumber",
+        label: "Course Number"
+    },
+    {
+        key: "courseTitle",
+        label: "Course Name"
+    },
+    {
+        key: "instructorName",
+        label: "Professor"
+    },
+    {
+        key: "gradeAverage",
+        label: "Average Grade"
+    },
+    {
+        key: "gradeSpread",
+        label: "Grade Spread"
+    },
+    {
+        key: "gradedTotal",
+        label: "Total Graded"
+    },
+    {
+        key: "enrollmentTotal",
+        label: "Total Enrolled"
+    }
+]
+
+const Summary = () => {
+    const [cutoff, setCutoff] = useState(SUMMARY_DISPLAY_COUNT);
+    const [sortBy, setSortBy] = useState<keyof SummaryRow>("gradeAverage");
+    const [sortReverse, setSortReverse] = useState(true);
+    const [displaySummaries, setDisplaySummaries] = useState(subjectSummaries.slice(0, cutoff));
+    const observerRef = useRef<HTMLDivElement>(null);
+
+    const onHeaderClick = (key: keyof SummaryRow) => {
+        if (sortBy === key) {
+            setSortReverse(!sortReverse);
+        } else {
+            setSortBy(key);
+            setSortReverse(false);
+        }
+        setCutoff(SUMMARY_DISPLAY_COUNT);
+    };
+
+    useEffect(() => {
+        const observer = new IntersectionObserver((entries) => {
+            if (entries[0].isIntersecting) {
+                setCutoff(c => c + SUMMARY_DISPLAY_COUNT);
+            }
+        });
+
+        const node = observerRef.current;
+
+        if (node) {
+            observer.observe(node);
+        }
+
+        return () => {
+            if (node) {
+                observer.unobserve(node);
+            }
+        }
+    }, [observerRef]);
+
+    useEffect(() => {
+        setDisplaySummaries(subjectSummaries.filter(row => row.gradedTotal !== 0).sort((a, b) => {
+            if (typeof a[sortBy] === "number" && typeof b[sortBy] === "number") {
+                return ((a[sortBy] as number) - (b[sortBy] as number)) * (sortReverse ? -1 : 1);
+            } else if (typeof a[sortBy] === "string" && typeof b[sortBy] === "string") {
+                return (a[sortBy] as string).localeCompare(b[sortBy] as string) * (sortReverse ? -1 : 1);
+            }
+            return 0;
+        }).slice(0, cutoff));
+    }, [cutoff, sortBy, sortReverse]);
```

Comment:
[don't think we need the effect](https://react.dev/learn/you-might-not-need-an-effect) here since it looks like this is just transforming data for rendering:

> You don’t need Effects to transform data for rendering. For example, let’s say you want to filter a list before displaying it. You might feel tempted to write an Effect that updates a state variable when the list changes. ([read more](https://react.dev/learn/you-might-not-need-an-effect))

also, small nit since I think the dataset is too small for it to really matter but if you're up for some optimization, it looks like as the user scrolls more and the cutoff increases, the size of the array that's being sorted grows and grows even though we've already done the work of sorting the top part of the array as long as the sorting key did not change (been a while since I've had to do algorithms but I think by the time the user scrolls to the end it ends up being O(n^2) rather than n log n) 

is there a way for us to take advantage of the fact that we've already done the work for the first part of the array and it's just the rest of the array that needs to be sorted as the user scrolls?